enable pressure coordinates for packages exf, seaice, ggl90

mjlosch opened the pull request:
## What changes does this PR introduce?
enable oceanic pressure coordinates with pkgs exf, seaice, ggl90 (including idemix)
and fix small issues related to pressure coordinates
## What is the current behaviour? 
pressure coordinates can only be used with the core model

## What is the new behaviour 
pressure coordinates can be used with sea ice (seaice/exf) and a mixed layer model (ggl90) and penetrating shortwave radiation

## Does this PR introduce a breaking change? 
No, previous results are unaffected

## Suggested addition to `tag-index`
adjust pkg/exf, seaice, ggl90 to be used with pressure coordinates
- in exf and seaice this mainly means ensuring the proper use of the surface index
 + for seaice surface pressure loading is not possible, add a flag sIceLoadFac (in "data", default=1.) to be able to turn it off also for z-coordinates
 + surface tilt term cannot use ETAN but must use diagnostic phiHydLow
- for ggl90 
 + all depth related lengths are scaled to be measured in meters, so that no parameter changes are necessary (as they still are for core parameters in namelist file "data")
 + k surface and bottom index need to be adjusted
 + computation of sigmaR requires some care in do_oceanic_physics.F
- fix treatment of Qsw in apply_forcing.F for pressure coordinates
- read geopotential anomaly directly into phi0surf instead of via pLoad (which was a hack), specified with new filename variable geoPotAnomFile in data&PARM05
- adjust MXLDEPTH diagnostic for pressure coordinates
- add verification experiment as part of global_ocean.cs32x15
- add a python-script gendata.py to tutorial_global_oce_in_p that shows how initial conditions and geopotential anomaly file for pressure coordinate simulation are created (unfortunately, they are not the same as in this reference experiment)

mjlosch commented:
The promised reference experiment is a draft so far. In order to run it, one need to generate the input files with the provide gendata.py. So far there are no reference results.

mjlosch commented:
The test fails, because there is the extra verification experiment in global_ocean.cs32x15 without a corresponding file in the results directory. At this stage I do not want to have a full experiment, so I am going to ignore this failing test.

mjlosch commented:
@jm-c should we try and add GMredi to the packages available with pressure coordinates in the PR?

mjlosch commented:
Now tutorial_global_oce_in_p fails and I have no idea why. I cannot reproduce this on my linux box. Can I rerun the travis tests for previous commits (e.g. for this one ce9582a)?

mjlosch commented:
> Now tutorial_global_oce_in_p fails and I have no idea why. I cannot reproduce this on my linux box. Can I rerun the travis tests for previous commits (e.g. for this one [ce9582a](https://github.com/MITgcm/MITgcm/commit/ce9582a495c4a7c4658e15b96da1745e2aa608a2))?

The reason for this is https://github.com/MITgcm/MITgcm/pull/334/commits/dd74553bf3e4d4e49ed4880845734c993b194833 which changes the computation of `sigmaR` for pressure coordinates, which will then have an effect of vertical mixing by "implicit diffusion" in this experiment (not sure where the original commit message went, `git log model/src/do_oceanic_physics.F` shows it). 

The change in `sigmaR` computation is necessary to maintain symmetry between z and p-coordinates, so I think that this is a bug fix and since I am only talking to myself I will go ahead and update the reference experiment.

jm-c commented:
@mjlosch I merged master (from after merging PR #493 and before merging PR #492) into this branch. There was few (5) files with merge-conflicts, 4 were obvious to resolve but the file ggl90_calc.F was tricky and would be good to check that it done correctly, both the pcoord modif and the autodiff changes.
I am going to merge the latest master in (and fix an other merge-conflict there) so no need to change files in this branch until it's done.
Regarding forward testreport output, everything is good, no changes. But for the TAF adjoint test, it's not that great: many more re-computation warning and global_ocean.cs32x15 (all 4) + lab_sea (all 3) are failing. Note that it was also the case before I merged in the recent master branch.

mjlosch reviewed (commented):
Note to myself: `phiHydLow` is calculated within `calc_phi_hyd`, which is called from `dynamics`. Here, we use the `phiHydLow` of the previous timestep. Either, we re-compute `phiHydLow` (only if `usingPcoords=T` and `useSEAICE=T`), in each timestep before seaice_model is called, or we calculate `phiHydLow` once after `ini_fields`. The latter will fix a potential restart problem, and the former will in addition probably avoid storing phiHydLow of TAF. Not sure, what's better.

mjlosch commented:
We need to decide, if we want pressure coordinates to work with AD-code. The AD-code of `ggl90_calc` becomes very complicated because of the "dynamic" loop range and directions to that we have to do "unnecessary" stores just because TAF cannot know the direction of the loop (this is similar to convective_adjustment). I have now suppressed all recomputation warnings, but there's still a "potential conflict" warning, which I don't know how important that is (nor how to fix it). These warnings only show up in `verification_other/global_oce_*` (where I also noticed similar warnings related to salt_plume).

@jahn how do I fix this error related to `verification_parser.py`? All tests pass, but I still get a fail because I now have too many subtests:
```
Traceback (most recent call last):
  File "verification_parser.py", line 103, in <module>
    verification_parser(**vars(args))
  File "verification_parser.py", line 27, in verification_parser
    raise ValueError(error_message)
ValueError: 5 values given for threshold, but 6 subtests found.
Error: Process completed with exit code 1.
```

jm-c commented:
@mjlosch I am starting to look at code changes, will provide some comments regarding model/src hopefully soon.
I the mean time: 
1) AD testreport: seems that isomip & offline_exf_seaice are now getting more recomputation warnings (7 vs 4 and 35 vs 33).
   could you confirm ?
2) in global_ocean.cs32x15.in_p, I would prefer to have input bin files that don't share the same name when they are different (and they are): I would suggest something like:
     bathy_Hmin50.bin  --> bathy_in_P.bin
     lev_S_cs_15k.bin  --> lev_S_cs_flip15.bin
     lev_T_cs_15k.bin  --> lev_T_cs_flip15.bin

jahn commented:
> @jahn how do I fix this error related to `verification_parser.py`? All tests pass, but I still get a fail because I now have too many subtests:

The expected digits are set in .github/workflows/build_testing.yml.  I believe you want to insert a '10' in the third place for global_ocean.cs32x15 since the new subexp only has 10 digits agreement.

mjlosch commented:
@jm-c
1. I can confirm that, phi0surf and phiHydLow are new problems. I'll have a look at that.
2. I agree. Is there a simple way to rename these files?

mjlosch commented:
@jm-c
I fixed the (benign) phi0surf problem and will push that. The phiHydLow problem is also benign in the sense that TAF creates code to deal with this, but it has to do with the fact, that in this branch phiHydLow is exchanged in seaice_model.F if usingPcoords=T.
Rather than fixing the recomputation warning, which is just a symptom, I'd rather fix the cause, which is related to the problem that phiHydLow is computed in dynamics/calc_phi_hyd and never exchanged anywhere. It should probably be exchanged in do_fields_blocking_exchanges along with totPhiHyd (there's a weird store of the 2D phiHydLow within the k-loop of dynamics.F). And I couldn't find phiHydLow in any pickup scheme.

jm-c commented:
>     2\. I agree. Is there a simple way to rename these files?

If you use "git mv bathy_Hmin50.bin  bathy_in_P.bin" in global_ocean.cs32x15/input.in_p (+ similar for the 2 others) and then commit and push, this will keep the history.

jm-c commented:
@mjlosch regarding your recent changes and AD testreport: this fixes isomip & offline_exf_seaice additional recomps. But for some reasons, I now get more recomp. warnings in 1D_ocean_ice_column (7 vs 5 before) related to:
TAF WARNING  triggered by      : uice,vice
in do_oceanic_phys.
However, I sometimes got things wrong so it would be good for you to check first that I am not making this up.

jm-c commented on model/src/apply_forcing.F:
```diff
@@ -642,14 +642,23 @@ SUBROUTINE APPLY_FORCING_T(
       IF (linFSConserveTr) THEN
-       DO j=0,sNy+1
-        DO i=0,sNx+1
+       IF ( usingZCoords ) THEN
```
I think the original code here for linFSConserveTr is fine both for Z and P coords (+ the same in APPLY_FORCING_S).
Therefore I would recommend not to change it.

mjlosch replied:
OK, I guess, I don't understand this correction term. If it is supposed to be added at the free surface, then my code is wrong. I assumed it was supposed to be added at the surface of the ocean. I'll revert that.

jm-c commented on model/src/calc_wsurf_tr.F:
```diff
@@ -62,19 +62,35 @@ SUBROUTINE CALC_WSURF_TR(thetaFld, saltFld, wVelFld,
+           ENDDO
           ENDDO
-         ENDDO
+         ELSEIF (usingPCoords) THEN
```
I still think that the original code is fine for both Z and P coords ; and I have the impression that this additional piece of code is not right, because the free-surface is at "kSurfC" and not at kLowC, even with P-coords.

mjlosch replied:
Same as before. I don't remember, why I changed this (16 months ago). I was probably just looking for `kSurfC` and adjusting the code (in the wrong way). I don't think that I ever tested that.

mjlosch commented:
> @mjlosch regarding your recent changes and AD testreport: this fixes isomip & offline_exf_seaice additional recomps. But for some reasons, I now get more recomp. warnings in 1D_ocean_ice_column (7 vs 5 before) related to:

That's correct, I can reproduce that. I could easily fix that (one line), but I'd prefer to assess and merge PR #496 which addresses this in a more substantial way first, and then merge it into this PR. I have the feeling that this pcoord_fixes PR still requires a few days of work and PR #496 (because of the testreport part) would be very helpful here.

mjlosch commented:
@jm-c since the AD of ggl90 is so delicate, I suggest to add it to one of the regularity tested experiments (just compiler and generate the AD-code). Currently it's only tested in verification_other and in OpenAD/code_oad. Does it make sense to add it also to OpenAD/code_oad/packages.conf? Or would you rather have another AD test in global_ocean.cs32x15?

mjlosch commented:
I can confirm that with the merge 9191ca7, the extra recomputations in 1D_ocean_ice_column go away.

jm-c commented on model/src/apply_forcing.F:
```diff
@@ -664,30 +664,54 @@ SUBROUTINE APPLY_FORCING_T(
+     &                         -swfracb(2)*maskC(i,j,kp1, bi,bj))
+     &        *recip_Cp*mass2rUnit * gravitySign
+     &        *recip_drF(k)*_recip_hFacC(i,j,k,bi,bj)
         ENDDO
```
I am not sure about the maskC here when using P-coord since it is meant to be the mask at the level below the corresponding interface (rF(k) and rF(k+1)). So It would make more sense to have swfracb(1)*maskC(k-1) and swfracb(2)*maskC(k).

mjlosch replied:
Nice catch. I think you are right. I'll introduce some `kUp` and `kDown` and set them accordingly.

mjlosch replied:
It certainly looks more elegant! To go even further, one could replace lines 703/4 with just
`IF ( k.EQ.km ) swfracb(2) = 0. _d 0`

fix spelling of "depth" in the comment?

jm-c replied:
this is a good suggestion. I will push this soon. Thanks.

jm-c commented:
@mjlosch I added the "adjoint" label so that the CI check all OpenAD experiments. I am also going to make little changes
to the penetrating SW and after you can decide to keep this or to revert to your latest version.

jm-c commented:
Regarding OpenAD tests, it seems that the experiment that fails to compile, i.e., OpenAD, is related to pkg/ggl90:
If I remove this pkg from code_oad/packages.conf then I can compile and run and it pass the test (except the ggl90 one)

mjlosch commented:
My observation is that since the merge 19 days ago, my llc270 simulations have more variance ssh and bottom pressure variance. I have to look into that. This is also most likely related to ggl90. Something happened during the merge. Can you compile the OpenAD tests with code prior to this merge, i.e with commit cf804f7 ?

jm-c commented on model/src/forcing_surf_relax.F:
```diff
@@ -192,8 +192,8 @@ SUBROUTINE FORCING_SURF_RELAX(
-         WRITE(msgBuf,'(A,E24.17)')
-     &        'rm Global mean of SSTrelax =',  globAver
+         WRITE(msgBuf,'(A,e24.17,1X,E24.17)')
+     &        'rm Global mean of SSTrelax =',  myTime, globAver
```
I think it's useful to have time or iteration written on same line as global mean. However, as it is now one might wonder why there are 2 printed numbers (with same number of digits) like this:
(PID.TID 0000.0001) rm Global mean of SSSrelax = 0.21600000000000000E+05 -0.12746283611881477E-08
For clarity (+ my own preference for iter number), I would print something like:
(PID.TID 0000.0001) rm Global mean of SSSrelax =  -1.27462836118815E-09 @ it=         6
which I get from:
         WRITE(msgBuf,'(A,1PE23.14,A,I10)')
     &        'rm Global mean of SSSrelax =', globAver, ' @ it=', myIter

jm-c replied:
Note that blanks were not preserved in my previous comment (although I put 10 of them between "it=" and "6").

mjlosch replied:
I agree, and I also agree that we can drop this change for now and have a separate PR that fixes more issues in this file.

jm-c replied:
OK, I will push this little change in forcing_surf_relax.F and will open a new PR to improve remove_mean.F with a similar improved print statement.

jm-c commented on model/src/calc_oce_mxlayer.F:
```diff
@@ -91,36 +105,39 @@ SUBROUTINE CALC_OCE_MXLAYER(
            rhoMxL(i,j) = rhoSurf(i,j)
      &                 + MAX( rhoMxL(i,j)*hMixCriteria, dRhoSmall )
-           hMixLayer(i,j,bi,bj) = rF(1)-R_low(i,j,bi,bj)
+           hMixLayer(i,j,bi,bj) = ( rF(kSrf+deltaK)-R_low(i,j,bi,bj) )
```
there is a problem here: with method 1 default hMixLayer is set to full column depth (in case criteria is never met). This was and is still the case for Z-coord but for P-coord R_low is the top pressure and is equal to rF(Nr+1) so that the default is zero.

jm-c commented on model/src/calc_oce_mxlayer.F:
```diff
@@ -91,36 +105,39 @@ SUBROUTINE CALC_OCE_MXLAYER(
           DO i=1-Olx,sNx+Olx
-           IF ( k.LE.klowC(i,j,bi,bj) .AND.
-     &          rhoLoc(i,j).GE.rhoMxL(i,j) ) THEN
+           kk = k.LT.klowC(i,j,bi,bj).AND.k.GE.kSurfC(i,j,bi,bj)
```
the symetric way to re-write line 115 should be kk=k.LE.klowC .AND. k.GE.kSurf

jm-c commented on model/src/calc_oce_mxlayer.F:
```diff
@@ -135,45 +152,54 @@ SUBROUTINE CALC_OCE_MXLAYER(
-            rhoMxL(i,j) = -1.
-           ENDIF
+          IF ( klowC(i,j,bi,bj) .GT. 0 ) THEN
+           hMixLayer(i,j,bi,bj) = drF(kSrf) * maskC(i,j,kSrf,bi,bj)
```
There was no maskC before and I don't think we need to add one. Same thing 4 lines below.

jm-c commented on model/src/calc_oce_mxlayer.F:
```diff
@@ -135,45 +152,54 @@ SUBROUTINE CALC_OCE_MXLAYER(
 C     (from surface down to here) by factor hMixCriteria
 
+        IF ( usingPCoords ) THEN
+         kTop    = Nr-1
```
2 things: 1) since kTop & kBottom depend on which method is used, would be cleaner to set them, for each method, within the corresponding method IF block.
2) for method 2, kBottom should be set to 2 (nstead of 1 currently) to make it symmetric.

jm-c commented:
@mjlosch 2 comments:
1) I made some changes in calc_oce_mxlayer.F that address some problems that I saw there. I might be wrong on few of them so you will need to check.
2) regarding the small changes in "remove_mean.F", may be we could just drop them and keep this file unchanged for now. There are some issues about these 2 routines (see issue #501) and we could import your changes from this file in the future PR that will improve  remove_mean.F

jm-c reviewed (changes requested):
It's a partial review of just pkg/ggl90 changes. I took a look at the 2 files with most changes, i.e.,
ggl90_calc.F and ggl90_idemix.F but I am a bit lost when I want to know where 3-D variables are located, at level interface (where wVel is) or level center (where T & S are). I think it would help if this kind of information was added in the code at the place where these 3-D var are declared.
Regarding IdeMix variables, the description of parameters and variables seems missing in GGL90.h, so it dos not help.
- using depth variable time-step dTtracerLev(k) instead of a single time-step: I am not sure if it makes much sense and, in case this is used for reaching an equilibrium tracer distribution, if it would still be better to use a single time-step. The other thing is that dTtracerLev(k) is at level center and I thought GGL90TKE was at level interface, so I am confused.
- mxlMaxFlag= 2 and mxlMaxFlag=3 code are a bit tricky since GGL90mixingLength is updated according to value at level below and above and I am not sure that the P-coord version is right: for example, line 402-403, for k=kTop (=Nr), km1= Nr and GGL90mixingLength is always increased by drF(Nr). In original Z-coord code, the kBot level is done separately from the 2nd k loop and I suspect that for the P-coord version the kTop level should be done separately from the first k loop. But to be sure, I would suggest to write explicitly a P-coord version next to the original Z-coord version (for both mxlMaxFlag=2 & 3) with IF ( usingPCoords ) THEN/ ELSE/ ENDIF.
this will be easy to check (and will very likely fix the OpenAD compilation issue) and from there, could try to combine the 2 coord version with clever k loop (but we will be able to check that results don't change).
- I ahve the impression that all maskC(k) in ALLOW_GGL90_SMOOTH blok sould be changed to maskC(k+delK).
- some simplifications that could help: a) there are MIN(2,Nr) in ggl90_calc.F ; I don't think it make much sense to use this pkg with Nr=1 so could we just skip this S/R call if Nr = 1 ?
b)  with mxlMaxFlag=0,1,2 : no need to compute rMixingLength 2 times.
c) line 812 & 889: we don't need a MAX in kBottom=MIN(MAX(kSurfC(i,j,bi,bj)+1,2),Nr), and kBottom=MIN(kSurfC(i,j,bi,bj)+1,Nr) is simpler

mjlosch commented:
thanks for this very helpful review, @jm-c 
I can add more description in GGL90.h. For the IDEMIX variables this may be difficult, because I did not write this code, but just got it from Carsten Eden and checked it in. I am assuming that the 3D variable are generally also defined at the cell interfaces, but I'll have to check that, also the exact meaning of the fields ... requires a bit of time.

I'll go through the other comments carefully.

Your changes to calc_oce_mxlayer.F seem to have fixed this diagnostics (in my runs it is just that). I also found that my llc270 simulations in z and p are, contrary to what I said before, no more different between each other that before the merge 715125c, which is nice. Not clear why I seem to have more variability than before, but that may be a problem in my analysis.

jm-c commented:
@mjlosch regarding latest changes in calc_oce_mxlayer.F : I did not check which line is involved, but latest version allows both global_ocean.gm_k3d & global_ocean.gm_res (from verification_other) to pass with no change compared to master whereas with previous version both were failing (only 4 and 3 digits match for cg2d). Note that these 2 experiments use the mixed-layer depth inside GM-Redi K-3D.

mjlosch commented:
> @mjlosch regarding latest changes in calc_oce_mxlayer.F : I did not check which line is involved, but latest version allows both global_ocean.gm_k3d & global_ocean.gm_res (from verification_other) to pass with no change compared to master whereas with previous version both were failing (only 4 and 3 digits match for cg2d). Note that these 2 experiments use the mixed-layer depth inside GM-Redi K-3D.

That's great, and also in line with my observation in my comparsions between z and p-coordinates. Thanks for fixing this S/R!

mjlosch commented:
@jm-c I have addressed some of your concerns and will push that code today:
1. I added more documentation about where the (local) variables a located. GGL90.h and Idemix documentation will follow.
2. I followed all of the simplification suggestions (last bullet). Currently I just return to the calling routine if Nr=1, but hide this from TAF/OpenAD. We may want to just not call ggl90_calc from do_oceanic_phys. What is your preference?
3. fixed the maskC/S/W index in the GGL90_SMOOTH block, but since it also affects maskS/W, this modification changes results. I would either ask you to update the results or I do it temporarily with my own "reference" machine.
4. I have not yet reverted to constant deltaT, but we can do that later.

about the 
> mxlMaxFlag= 2 and mxlMaxFlag=3 code

This is indeed tricky and you have put your finger into the problem (but since I have done this so long ago, I don't really remember what my conclusions were back then). Here's the problem: GGL90 variables live at the interfaces (w-points), just as the vertical diffusivity/viscosity coefficients. Values at k=1 and k=Nr+1 do not make any sense because there's nothing to mix into the atmosphere or into the bottom of the domain. Still the original code computes something for k=1, but obviously nothing for k=Nr+1, because the field is not defined. For z, k=1 is the "surface" interface and GGL90mixingLength(1) = GGL90mixingLengthMin. For p, the equivalent level is Nr+1, which is not defined, so I had to "fake" that level, but I think that this is fine because GGL90mixingLength = GGL90mixingLengthMin for all k before we start with anything; so that should be safe and no extra code is required to be symmetric. Alternatively on could define the 3D fields from 1 to Nr+1 (or 2 to Nr) and that way everything could be symmetric.
At the bottom (in z: k=Nr, in p: k=2, because k=1 is the level/interface below the center of the bottom cell of the domain, that for z is not defined), there's another `GGL90mixingLength(...,Nr) = MIN(GGL90mixingLength(...,Nr), GGL90mixingLengthMin+drF(Nr)`. In p-coordinates, this should be `GGL90mixingLength(...,2) = MIN(GGL90mixingLength(...,2), GGL90mixingLengthMin+drF(1)`, and I think, that's what the code does (but will check again).

Does that make sense? I will still follow your suggestions and separate p and z code. Should we actually push this here?

jm-c commented:
@mjlosch 2 comments: 1) your suggestions look good, please go for it and I will take a look. I think it's still worth, for now, to separate the p & z coord case for mxlMaxFlag= 2 & 3 : it's not a lot of code and will be easier to check + then we will gain confidence and can see later how this could be written in a smart way with loess lines of code.
2) regarding update of output.txt, I could do that on villon but would wait until mxlMaxFlag= 2  is fixed (since the new P-coord test "in_p" uses that).

mjlosch commented:
I have just followed your suggestion and came to the conclusion that the code was nearly correct (at least for mxlMaxFlag= 2 it was correct). I will now check in a version were I haven't done much but rename a few indices to make their meaning clear. After that we can split this for  mxlMaxFlag= 2 into separate sections (but not this week). For mxlMaxFlag= 3, it will be totally analogous. OK?

jm-c reviewed (commented):
Finishing my review:
I look at changes in pkg/exf and pkg/seaice and they seem good. I made minor adjustments (mainly to have more consistent local variable name) but preserved all coding logic and computation.
From my point of view, the remaining things are:
1) still few things to sort out with pkg/ggl90
2) restart with pkg/seaice dynamics in P-coord.

mjlosch commented:
@jm-c I have now a version of ggl90_calc.F, that preserves the results, combines p and z version, and has a lot fewer store directives. Should I push it?

jm-c commented:
@mjlosch yes please. But does it pass the OpenAD tests (that are currently OK) ? If not we might need to put back some pieces from current version.

mjlosch commented:
The failing OpenAD test is super-annoying. I'll revert to a version with separate p- and z- coordinate k-loops, this will mean an extra 6 store directives that will just generate extra memory requirements and not be used (because we can only use either p or z-coordinates, but we need store directives for both cases). At least I know now how to reduces this extra memory requirement a little, compared to 4a7c280. Alternatively, we could use ALLOW_OPENAD flags to exclude the new pcoord-code like:
```
#ifdef ALLOW_OPENAD
     DO k=2,Nr
#else
     DO k=kTop,kBot,kDir
#endif
```
what do you think, @jm-c ?

jm-c commented:
@mjlosch As I mentionned earlier, the k loops for  mxlMaxFlag=2 &3 are tricky since the range and direction (+1 or -1) are not hard coded (the AD compiler does not know) and  mxLenght(K) = fct(mxLength(k),mkLength(kk)) but the AD compiler does not know if/when kk is different or the same as k.
Are we sure that the TAF ad code with current version of ggl90_calc.F is right ?
If no, then we should probably revert to previous version. 
If yes, then we could use ALLOW_OPENAD ifdef to make a longer version just for OpenAD (but it will be a bit more complicated than:
#ifdef ALLOW_OPENAD
     DO k=2,Nr
#else
     DO k=kTop,kBot,kDir
#endif
since the OpenAD part will have IF usingZcoord/Pcoord).

mjlosch commented:
OK, I'll revert my merging of the p- and z-code. It's not the end of the world.

jm-c commented:
@mjlosch I took a look at latest version: the case  mxlMaxFlag=2 looks good. But for the case  mxlMaxFlag=3, I have 2 comments:
1) I don't find the where the lines 445-446 of the original (master) code are now.
2) in the original (master) code mxlLength was not set to a minimum of GGL90mixingLengthMin (or at least I don't see it) before computing tmpmlx but now it looks like it is.
It's possible that I missed something or that one or both don't matter, but since we are using mxlMaxFlag=3 in any of the verification experiment we would not know it this is different from original version.

mjlosch commented:
@jm-c I was not careful with mxlMaxFlag=3 and the code was wrong. Now I fixed the code and checked it with a modified version of vermix.ggl90; compared to master, there are no differences anymore. We should consider using mxlMaxFlag=3 in at least one of the verification experiments (either vermix.ggl90 or global_ocean.90x40x15.idemix, both which use mxlMaxFlag=2).

jm-c commented:
@mjlosch I think it would be good to have mxlMaxFlag=3 tested. For me it does not matter too much if this is in vermix or
global_ocean.90x50x15. The remaining question would be if we do this in this PR or in a specific short one (that could be merged in before this one). What would be your preference ?

mjlosch commented:
then I opt for vermix and short extra PR. And I let you do it, because you'd have to update the reference output anyway, right?

jm-c commented:
@mjlosch I took a look at latest ggl90_calc.F and it looks good to me too. I may make some minor edits to few modified files in pkg/ggl90 but I don't want to interfere with any of your potential/future changes so let me know when it's a good time.
And otherwise, the only thing remaining would be to fix the p-coord restart with pkg/seaice.

jm-c commented:
> then I opt for vermix and short extra PR. And I let you do it, because you'd have to update the reference output anyway, right?

I can do that. May be you could send me (via email) your updated vermix data.ggl90 ? Thanks.

mjlosch commented:
I can stop working on this now. I have not yet reverted the deltaT variable back to a constant local one. That's a small change and so I think you can do that along with your minor edits?

Restart: How do you want to proceed with this. Recompute phiHydLow after restart, or add phiHydLow as an extra field to the pickup? 
Related: there's now an exchange of phiHydLow at the beginning of seaice_model.F that should probably be moved to somewhere else.

jm-c commented:
For the restart, with current code, I don't think we have a choice: I we are using staggerTimeStep, recomputing phiHydLow after reading pickup file will not give a perfect restart. 
Note that we might change this if we find out that a lag of 1 time-step in phiHydLow for seaice dynamics is not very stable, but it would require to always (i.e., every time-step) recompute phiHydLow before seaice-dynamics when using staggerTimeStep.

jm-c commented on pkg/ggl90/ggl90_calc.F:
```diff
@@ -662,13 +771,15 @@ SUBROUTINE GGL90_CALC(
        km1=MAX(2,k-1)
        DO j=jMin,jMax
         DO i=iMin,iMax
+         IF ( usingPCoords) km1=MIN(Nr,MAX(kSurfC(i,j,bi,bj),k-1))
```
I have the impression that if should be MIN(Nr,MAX(kSurfC+1,k-1)), since for P-coord, kSurfC is the very bottom interface level (in Z it is kLowC+1).

mjlosch replied:
Let's see if I can get this right: `km1` is used to compute the interpolated `0.5*(kappaE(k)+kappaE(km1))` at  the cell centres, so in the interior `km1` needs to be `k-1`. `kappaE(k)` is defined at the interface between level `k` and `k-1`. 
For the bottom-most cell (`k=kSurfC`), `kappaE(k-1)` is at the interface between first and then second dry cell below the bottom interface.
 If `kSurfC=k`, it is because `hFacC(k)>0` (ini_grids_etc.F), but `hFacC(k-1)=0` (below `k`, for P) and `max(kSurfC=k,k-1)=k`. `recip_hfacc(k-1)=0.`, and everything is masked. For `k=kSurfC+1` (second level above bottom interface), `max(kSurfC,kSurfC)=k-1`, but `kappaE(k-1)` is at an interface below the bottom interface and should not be defined (or is zero), so that `kappaE=0.5*kappaE(k)`. At `k=kSurfC+2` things are OK again with `max(kSurfC,k-1=kSurf+1) = kSurf+1 = k-1`.
If we use your suggestion, then `max(kSurfC+1,k-1) = k+1`, at `k=kSurfC`, which appears strange at first, but it doesn't matter, because it is masked by `recip_hfacc(k-1)=0`. At `k=kSurfC+1` (second level above bottom interface), we'd have `max(kSurfC+1, k-1=kSurfC) = k`, so that interpolated `kappaE=kappaE(k)`, which is defined, and at `k=kSurf+2`, `max(kSurfC+1,kSurfC+1) = kSurfC+1 = k-1`, etc. So I think your suggestion makes sense.

It doesn't change my test results at all, though.
Let me know, when you are done and I will push a fix.

jm-c replied:
I am not editing ggl90_calc.F anymore. However, once PR #506 is merged in, I will update this branch (to get vermix to test mxlMaxFlag=3).

jm-c replied:
One reason it does not change results in cs32.in_p might be that near the bottom the TKE is close to it's minimum value (so early in the simulation) or may be with such coarse vertical resolution it not sensitive to near bottom flux of TKE ?

mjlosch replied:
Not sure about that. My test (based on tutorial_global_in_p) runs for 100 years (deltaTtracer=1d) with ggl90. I get differences between p and z, but TKE near the bottom is non-zero in many places, see attached plot for a section though the Atlantic, and I get zero difference with or without the `+1` in this line.
![image](https://user-images.githubusercontent.com/30285667/127872991-a263351f-6f17-46f6-a991-1c38bbf267c8.png)

jm-c commented:
@mjlosch I made few changes to ggl90_calc.F, the main one is related to the use of maskC with ALLOW_AUTODIFF in SQRTTKE calculation. Could you check and confirm that my modification is correct ?
Also noticed a potential issue with km1 (I left a comment just now) but it does not have much effect on experiment cs32.in_p output.
And will take a second pass at ggl90_idemix.F changes (hopefully soon).

mjlosch commented:
@jm-c I am just going through the IDEMIX parameter descriptions. I didn't really adjust Carsten's code very much, back when I added this to the repository, so there's a lot of undocumented code, and variables. E.g. IDEMIX_V0, does not need to be a global variable, as far as I can see. I suggest that we will focus only on the p-coord-related stuff and I'll create a different PR, where I clean up the code, and maybe adjust the defaults to more recent values.

mjlosch commented:
> I made few changes to ggl90_calc.F, the main one is related to the use of maskC with ALLOW_AUTODIFF in SQRTTKE calculation. Could you check and confirm that my modification is correct ?

This look good to me.

jm-c commented:
> @jm-c I am just going through the IDEMIX parameter descriptions. I didn't really adjust Carsten's code very much, back when I added this to the repository, so there's a lot of undocumented code, and variables. E.g. IDEMIX_V0, does not need to be a global variable, as far as I can see. I suggest that we will focus only on the p-coord-related stuff and I'll create a different PR, where I clean up the code, and maybe adjust the defaults to more recent values.

Sounds good.

mjlosch commented:
The IDEMIX code appears to be broken. At least the differences between p and z-coordinates are much larger than without IDEMIX:
![image](https://user-images.githubusercontent.com/30285667/127975729-ee1eb3c0-1089-4fc7-aafd-2784fb79b24d.png)

@jm-c Unless you have already found some obvious bugs, I think it would make more sense to address the restart problem first (I think it is easier for you than for me to add this extra field to the pickup files), while I try to fiddle with IDEMIX (or we decide to leave IDEMIX entirely to a dedicated PR to get this one merged earlier).

jm-c commented:
@mjlosch I found some masking issues in ggl90_idemix.F for P-coord and when I make changes there it affects results of cs32.in_p new test. Will probably push something later today (but not right now since I have other things to do). 
Regarding restart, I could do that (I have done similar addition recently) but if you want or if you already started you are welcome to go/continue.

mjlosch commented:
I have not done anything on the restart problem, yet. But I am very much interested in the masking issues and if they affect my test, too (in a positive way).

mjlosch commented on pkg/ggl90/ggl90_idemix.F:
```diff
@@ -110,9 +133,9 @@ SUBROUTINE GGL90_IDEMIX(
-         Nsquare(i,j,k)= max( 100.*fxb*fxb, Nsquare(i,j,k) )
+     &                  * sigmaR(i,j,k) * coordFac
+         fxb = MAX( 1. _d -6, ABS( fCori(i,j,bi,bj) ))
+         Nsquare(i,j,k)= MAX( 100.*fxb*fxb, Nsquare(i,j,k) )
```
Here we could have instead: 
```
          fxb = MAX( 1. _d -12, fCori(i,j,bi,bj)*fCori(i,j,bi,bj) )
          Nsquare(i,j,k)= MAX( 100.*fxb, Nsquare(i,j,k) )
```

mjlosch commented on pkg/ggl90/ggl90_idemix.F:
```diff
@@ -154,12 +177,12 @@ SUBROUTINE GGL90_IDEMIX(
          DO i=1-OLx,sNx+OLx
-          fxb = 0.5*min( _dxF(i,j,bi,bj), _dyF(i,j,bi,bj) )*fxa
-          IDEMIX_V0(i,j,k,bi,bj) = min( IDEMIX_V0(i,j,k,bi,bj), fxb )
+          fxa = SQRT( 1. _d 0/( deltaTloc * IDEMIX_tau_h ) )
```
If deltaTloc = constant with depth again, we can move this line out if the k-loop again.

mjlosch reviewed (commented):
I just found two small things in ggl90_idemix, we may want to remember to adjust.

jm-c commented:
@mjlosh I pushed my changes to ggl90_calc.F (+ see details of the commit message). Let me know what you think of these.

mjlosch commented:
@jm-c thanks for the explanation. I think your changes make a lot of sense, and could also be applied elsewhere in pkg/ggl90, where instead mask(k)*mask(k-1) is used. These changes don't fix the differences between p and z runs with IDEMIX. 

Instead I traced the problem back to a division by `MAX(Nsquare,GGL90eps)`, with `GGL90eps=2.23e-16` (line 294 in ggl90_calc.F) when GGL90mixingLength is computed. As Nsquare depends on sigmaR, this regularisation is too weak to even out small differences, which then lead to significant differences in mixing length etc. replacing GGL90eps with 1e-12 reduces the differences between p and z-coordinates. 

I am not sure if this PR request is the right place to fix this (one could use a larger fixed value or a runtime parameter instead of GGL90eps).

jm-c commented:
> @jm-c thanks for the explanation. I think your changes make a lot of sense, and could also be applied elsewhere in pkg/ggl90, where instead mask(k)*mask(k-1) is used. These changes don't fix the differences between p and z runs with IDEMIX.
> 
> Instead I traced the problem back to a division by `MAX(Nsquare,GGL90eps)`, with `GGL90eps=2.23e-16` (line 294 in ggl90_calc.F) when GGL90mixingLength is computed. As Nsquare depends on sigmaR, this regularisation is too weak to even out small differences, which then lead to significant differences in mixing length etc. replacing GGL90eps with 1e-12 reduces the differences between p and z-coordinates.
> 
> I am not sure if this PR request is the right place to fix this (one could use a larger fixed value or a runtime parameter instead of GGL90eps).

If this changes output of experiments that are not using P-coord, it would be better, I think, to make this change in an other PR. This is just my opinion.

jm-c commented:
@mjlosch If you are OK with the changes I made in idemix code then I will update the reference output for the new test cs32.in_p.

mjlosch commented:
> @mjlosch If you are OK with the changes I made in idemix code then I will update the reference output for the new test cs32.in_p.

@jm-c Yes, I am OK. Please go ahead.

jm-c commented:
I added phiHydLow to pickup file and it does fix the restart test when using pkg/seaice in P-coord. However, it seems that the new test experiment cs32.in_p is still not passing the restart test, but now it does when I switched off idemix (useIDEMIX=F in data.ggl90).

mjlosch commented:
> I added phiHydLow to pickup file and it does fix the restart test when using pkg/seaice in P-coord. However, it seems that the new test experiment cs32.in_p is still not passing the restart test, but now it does when I switched off idemix (useIDEMIX=F in data.ggl90).

@jm-c I have narrowed this down to the horizontal diffusion terms. When I set `IDEMIX_tau_h=0.`  (default is 10 days), the restart test passes. I have tried all sorts of things, but I have no idea, why this is so. I have commented out everything so that the diffusion contribution is computed only from constant grid parameters (dx,dy,dr, maskW/S/C, deltaTloc), but that still does not work. Do you have any idea, what that could be related to? It's really only a few lines of code.

jm-c commented:
@mjlosch Did you try a similar set-up (cs32 with ggl90 with idemix, z*, seaice) in Z-coord to check if restart pass ? this might be useful to know (and may be fix in an other PR if it turns to be unrelated to P-coord).

mjlosch commented:
Good point. I tried with z*-coords (I used input.seaice, although it has nothing to do with seaice, restart also fails with useSEAICE=.FALSE.) + ggl90+idemix and the problem is reproducible, i.e. it seems to be specific to the combination of cs32 with ggl90 and idemix (I turned off the nonlinear free surface and the problem persists). Really puzzling, but it has little to do with p-coords, so I agree, let's defer it to a different PR.

jm-c commented:
We could start an issue with "to-do-list" things about idemix. There was 1 global array that could be made local. Not sure what the EXCH of IDEMIX_V0 is doing neither. And could take a look later on about this restart problem as I think I know what I would like to check.

jm-c commented:
@mjlosh I have a hack that fix the restart test for cs32.in_p ; we could chat on Monday how we proceed.

jm-c commented:
Just to summarize where we are and what's next:
1. few things to fix in idemix code (not related to P-coord but that show up in new experiment test) are listed in issue #511 and will be addressed in an other PR
2. pkg/gmredi updates for P-coord will go in different PR
3. all other things/comments have been addressed so we could merge this PR soon.
For practical reasons (several daily testreport test will be missing until Wed) we could delay the merging until Wed, with a bit more time if someone has comments to formulate.

jm-c commented:
> Just to summarize where we are and what's next:
> 
>     1. few things to fix in idemix code (not related to P-coord but that show up in new experiment test) are listed in issue [improve IDEMIX code #511](https://github.com/MITgcm/MITgcm/issues/511) and will be addressed in an other PR
> 
>     2. pkg/gmredi updates for P-coord will go in different PR
> 
>     3. all other things/comments have been addressed so we could merge this PR soon.
>        For practical reasons (several daily testreport test will be missing until Wed) we could delay the merging until Wed, with a bit more time if someone has comments to formulate.

Obviously, will have to fix latest commit that breaks the OpenAD test experiment with OpenAD.
